app/src/services/permit.ts

```diff
@@ -173,22 +134,79 @@ const service = {
+   * Upsert a Permit
    * @param {Permit} data Permit object
    * @returns {Promise<Permit | null>} The result of running the update operation
    */
-  updatePermit: async (data: Permit) => {
+  upsertPermit: async (data: Permit) => {
     try {
-      const response = await prisma.permit.update({
-        include: {
-          permit_type: true
-        },
-        data: { ...permit.toPrismaModel(data), updated_by: data.updatedBy },
-        where: {
-          permit_id: data.permitId
+      const permitDataWithId = { ...data, permitId: data.permitId || uuidv4() };
+
+      // Upsert the permit
+      const response = await prisma.$transaction(async (trx) => {
+        const response = await trx.permit.upsert({
+          include: {
+            permit_type: true
+          },
+          where: {
+            permit_id: permitDataWithId.permitId
+          },
+          update: { ...permit.toPrismaModel(permitDataWithId), updated_by: permitDataWithId.updatedBy },
+          create: {
+            ...permit.toPrismaModel({ ...permitDataWithId }),
+            created_by: permitDataWithId.createdBy,
+            updated_by: permitDataWithId.updatedBy
+          }
+        });
+
+        // Delete any permit tracking that is not in the new data
+        await trx.permit_tracking.deleteMany({
+          where: {
+            permitId: permitDataWithId.permitId,
+            permitTrackingId: {
+              notIn: permitDataWithId.permitTracking
+                ?.map((x: PermitTracking) => x.permitTrackingId)
+                .filter((x) => x) as number[]
+            }
+          }
+        });
+
+        // Upsert the permit tracking
+        if (permitDataWithId.permitTracking?.length) {
+          await Promise.all(
+            permitDataWithId.permitTracking.map(async (x: PermitTracking) => {
+              if (x.permitTrackingId) {
+                await trx.permit_tracking.update({
+                  where: {
+                    permitTrackingId: x.permitTrackingId
+                  },
+                  data: {
+                    permitId: permitDataWithId.permitId,
+                    trackingId: x.trackingId,
+                    sourceSystemKindId: x?.sourceSystemKindId,
+                    shownToProponent: x?.shownToProponent ?? false,
+                    updatedBy: permitDataWithId.updatedBy
+                  }
+                });
+              } else {
+                await trx.permit_tracking.create({
+                  data: {
+                    permitId: permitDataWithId.permitId,
+                    trackingId: x.trackingId,
+                    sourceSystemKindId: x?.sourceSystemKindId,
+                    shownToProponent: x?.shownToProponent ?? false,
+                    createdBy: permitDataWithId.createdBy,
+                    updatedBy: permitDataWithId.updatedBy
+                  }
+                });
+              }
+            })
+          );
         }
+        return response;
```

Comment:
Two observations:
- The three main blocks of operations may be await Promise.all wrapped together - I'm not seeing any immediate ordering constraints on the three operations (aka something must execute first as the next call is dependent on the result)
- The result of trx.permit.upsert is only used on the return - may be good idea to do this operation last, as then you could just return that action instead of declaring another const just to use it in one place.
There's an opportunity to streamline the async subfunctions here.